### DISCUSSION TITLE
Valetudo 2022.09.0

### DISCUSSION THREAD

**Discussion by Hypfer** [Hypfer/Valetudo#1571] [Release Discussions]
<div align="center">
    <img src="https://raw.githubusercontent.com/Hypfer/Valetudo/master/assets/logo/valetudo_logo_with_name.svg" width="800" alt="valetudo">
    <p align="center"><h2>2022.09.0</h2></p>
</div>

An interesting newly supported robot plus security and usability improvements

## Dreame P2148 Ultra Slim

[<img src="https://user-images.githubusercontent.com/974410/188237869-dfb5e033-1ca7-43e1-ab51-cce9019dc2b6.png" width=600>](https://user-images.githubusercontent.com/974410/188237869-dfb5e033-1ca7-43e1-ab51-cce9019dc2b6.png)

Thanks to generous user donations, I was able to buy a Mijia Ultra Slim Robot to take a look at.
As you can see, it is _very_ small, measuring only 5.5cm in height and 32.3cm in diameter.
For comparison:
The Z10 next to it measures a height of 9.6cm and a diameter of 35cm.

This makes the ultra slim a very interesting robot as it fills a niche that no other supported one did previously.
Apart from that, it's a decent vSLAM robot that - thanks to the ToF sensor in the front - doesn't bump into every wall that it can find.

Fortunately it turned out that even the latest available firmware is rootable via the very easy debug UART rooting method :)
Thus, there's now support for it in Valetudo.

Unfortunately however, it is a china-exclusive model, meaning that both it and any spare part have to be imported to wherever you live (unless you live in china that is).
Because of its size, it also uses completely different consumables than other dreames. Do keep that in mind if you're thinking about buying one.

Nonetheless, I am very happy to have it supported, as it is a proper solution to a problem that previously was unsolvable without outright replacing your furniture. There is simply nothing else like it in the market.
Thus, if you've ever wanted to use Valetudo but couldn't as all supported robots were too large for your scenario, then now there's an option for you.

I do want to stress again that this would not have been possible if it weren't for user donations.
There simply is no way to pay for all those robots out of our own pocket... especially considering the pricing of some upcoming models and even more so since we'll need at least two of them

[<img src="https://user-images.githubusercontent.com/974410/188237945-3c28cbce-99c9-4184-8d78-3da6230be434.png" width=600>](https://user-images.githubusercontent.com/974410/188237945-3c28cbce-99c9-4184-8d78-3da6230be434.png)

Thank you for your ongoing support!

## Encrypted Wi-Fi provisioning

[<img src="https://user-images.githubusercontent.com/974410/188238031-cdb7b422-a1f2-440c-973b-94c6104c604e.png" width=400>](https://user-images.githubusercontent.com/974410/188238031-cdb7b422-a1f2-440c-973b-94c6104c604e.png)

One thing that slightly annoyed me for a long time now was the fact that Wi-Fi provisioning was done via an unencrypted connection over an unencrypted Wi-Fi Access Point, meaning that anyone passively sniffing Wi-Fi traffic during the provisioning process could've in theory gained access to your Wi-Fi credentials.

To solve this unlikely scenario, the Valetudo frontend now requests an RSA-Key from the backend and encrypts the provisioning payload with it, before it is sent over the Wi-Fi. 
This slight improvement in security does not prevent MITM-attacks, however that wasn't ever the goal to begin with as those are even less plausible than the passive Wi-Fi sniffing attack.

If you want to know more about the reasoning behind this, feel free to take a look at [this very extensive code comment further explaining things.](https://github.com/Hypfer/Valetudo/blob/4ee0004b075abb7fcb274a3b4cd40bc9cd314cd7/backend/lib/webserver/capabilityRouters/WifiConfigurationCapabilityRouter.js#L25-L86)

### Larger ClientStructure hitboxes

[<img src="https://user-images.githubusercontent.com/974410/188238171-e88fbd75-8bf5-4ae6-8028-9d68e55a7fbd.png" width=350>](https://user-images.githubusercontent.com/974410/188238171-e88fbd75-8bf5-4ae6-8028-9d68e55a7fbd.png)

It should now be easier to interact with the map on mobile as the hitbox size of various things has been increased.

### Robot Coverage map

[<img src="https://user-images.githubusercontent.com/974410/188238369-6fb27a80-28a8-4fd3-906f-f8ddc9171b7a.png" width=600>](https://user-images.githubusercontent.com/974410/188238369-6fb27a80-28a8-4fd3-906f-f8ddc9171b7a.png)

Based on user feedback, the coverage map now renders the coverage and the robots' path.

### Map Data Export Button

[<img src="https://user-images.githubusercontent.com/974410/188238429-ca220d6f-658f-4844-baaf-fed72465365b.png" width=600>](https://user-images.githubusercontent.com/974410/188238429-ca220d6f-658f-4844-baaf-fed72465365b.png)

There's now a button that allows the user to download a ValetudoMap JSON e.g., for use with the [ValetudoMap to Minecraft Map converter.](https://github.com/Hypfer/Valetudo-Minecraft-Mapper)
Previously, users had to manually call the REST API route and save the JSON.

### MQTT resubscribe workaround

In some setups with MQTT brokers that don't have any persistence enabled, a broker restart or a loss of connectivity could lead to Valetudo not reacting to any MQTT commands anymore as all subscriptions were gone. It would still continue publishing status updates though, leading to even more confusion.

This issue should now be solved.

### Swagger UI

The Swagger UI now only displays routes that are supported by the currently running ValetudoRobot implementation.

## Quality improvements

[<img src="https://user-images.githubusercontent.com/974410/188238471-5d37529c-7ee3-4972-81bf-c410a562aa06.png" width=350>](https://user-images.githubusercontent.com/974410/188238471-5d37529c-7ee3-4972-81bf-c410a562aa06.png)

- Errors have been refactored to have a proper place in the abstraction
- The updater has been refactored
- Binaries are now 1MB smaller due to the removal of irrelevant sourcemaps

## The usual

[<img src="https://user-images.githubusercontent.com/974410/188238522-7a25b8dc-c536-456b-9eef-2cd6c7fe7916.png" width=350>](https://user-images.githubusercontent.com/974410/188238522-7a25b8dc-c536-456b-9eef-2cd6c7fe7916.png)

If you want to see Valetudo on more robots and/or like this release, you might want to consider donating if you haven't done so already:

[https://github.com/sponsors/Hypfer](https://github.com/sponsors/Hypfer)

[https://builder.dontvacuum.me/donations.txt](https://builder.dontvacuum.me/donations.txt)

### Features

- **ui**: ValetudoMap data export button [`f1aea85`](http://github.com/Hypfer/Valetudo/commit/f1aea85eaf54d25fbefefb745013e172e74fa92e)
- **vendor.dreame**: Add support for the P2148 Xiaomi Mijia Ultra Slim [`22419f4`](http://github.com/Hypfer/Valetudo/commit/22419f466df7a02580098738098a1b0acd900085)
- Encrypted Wi-Fi provisioning [`2b6dd04`](http://github.com/Hypfer/Valetudo/commit/2b6dd045aecb05e63303535763fff5a9dbf71f99)
- Classify robot errors [`e8715cb`](http://github.com/Hypfer/Valetudo/commit/e8715cb8b1005faa87a52e0c0f009d7baf8b7fd8)
- **vendor.dreame**: Add mop dock error code mapping [`6179e19`](http://github.com/Hypfer/Valetudo/commit/6179e191da8381a84a7fe7e720859a16b62c3ed4)
- **webserver**: Filter openapi spec to only display endpoints supported by the implementation in swagger ui [`3ab4033`](http://github.com/Hypfer/Valetudo/commit/3ab4033fbb135f736b203cb82f7fd26f0c3dddd9)
- **ui**: Improve coverage view [`4ac2928`](http://github.com/Hypfer/Valetudo/commit/4ac292862561713f93386944f12e850a4e1a5517)
- **ui**: Improve map interaction by allowing for slightly misaligned clicks/taps [`6e87ba3`](http://github.com/Hypfer/Valetudo/commit/6e87ba35072cf50a29846f09f9600400fdc6a31d)

### Fixes

- **mqtt**: Fix resubscribe on broker restart [`be33e12`](http://github.com/Hypfer/Valetudo/commit/be33e12148c81353c41dfd7c2d374263efe89a1f)
- **vendor.dreame**: Remove confusing MCU version logline [`4c21394`](http://github.com/Hypfer/Valetudo/commit/4c21394ed93a70887aff5e8be4aab04a3cd0cc0c)
- **miio**: Correctly update our stamp [`ee6a60d`](http://github.com/Hypfer/Valetudo/commit/ee6a60dfc5b94744faa641d953240ced52e8aa72)
- **vendor.dreame**: Remove broken mop only quirk [`dbbd0f0`](http://github.com/Hypfer/Valetudo/commit/dbbd0f0c3d847bafad6f3d3d12959693a675b022)
- **vendor.dreame**: Fix wrong quirk description [`afac067`](http://github.com/Hypfer/Valetudo/commit/afac0675af15ce1f8e457e5ed51ac835bdd71328)
- **vendor.dreame**: Ignore another useless piid on the 1C [`922aece`](http://github.com/Hypfer/Valetudo/commit/922aecef9525f8e82e7b8ab9ebe2596562e030c0)
- http.ServerResponse.socket can occasionally be null [`03d22dc`](http://github.com/Hypfer/Valetudo/commit/03d22dc6410ed043a2b95856408b72407ce08ad6)

### Refactoring

- **miio**: Code cleanup [`81809f1`](http://github.com/Hypfer/Valetudo/commit/81809f13007ae49045d1d1475d36629a59ea456d)
- **updater**: Break updater up into multiple files [`b9bcbe3`](http://github.com/Hypfer/Valetudo/commit/b9bcbe34c946a8b4f00abb58ddab29708b99107b)
- **vendor.dreame**: Change implementation name to something pronounceable [`b198b10`](http://github.com/Hypfer/Valetudo/commit/b198b106bcc042fb06502a69cf99861754304a28)
- **ui**: Reduce duplicate code in map structures and improve readability [`4f05dcd`](http://github.com/Hypfer/Valetudo/commit/4f05dcdd51c0e0ff4eb20e4f976d6704727a6803)

### Chores

- **release**: 2022.09.0 [`8e3c889`](http://github.com/Hypfer/Valetudo/commit/8e3c889735eb020912f3581d5b91f651daa8c0e8)
- remove body-parser in favor of built-in feature (#1568) [`4ee0004`](http://github.com/Hypfer/Valetudo/commit/4ee0004b075abb7fcb274a3b4cd40bc9cd314cd7)
- Misc promise handling cleanup [`278ba6e`](http://github.com/Hypfer/Valetudo/commit/278ba6e1af2899da78194aab4315fdfc84cca2a5)
- Remove obsolete code + comments [`7de1590`](http://github.com/Hypfer/Valetudo/commit/7de15906e785e8e9140107b9497da5d1e7d4726e)
- **ui**: Disable source maps for production builds to save 1mb compressed total binary size [`2a11a82`](http://github.com/Hypfer/Valetudo/commit/2a11a82424f1de0a9520093df0819d8ca2506cba)
- Minor cleanup [`e0a0800`](http://github.com/Hypfer/Valetudo/commit/e0a08001c0212cd0e8ef73b538c5de425ecdd27b)

<hr /><em>This discussion was created from the release <a href='https://github.com/Hypfer/Valetudo/releases/tag/2022.09.0'>Valetudo 2022.09.0</a>.</em>

**Comment by mrwsl** [comment]
So wait: did you actually buy a w10 pro? :tada:
